### ISSUE TITLE
ALSA snd_pcm_poll_descriptors_revents called with only a single descriptor 

### ISSUE THREAD

**Issue by timothyjward** [MusicPlayerDaemon/MPD#1204]
I've been working to build an ALSA output PCM, and I've found some ALSA usage in MPD which I think violates the expected usage of the ALSA API. Specifically in a call to `snd_pcm_poll_descriptors_revents`.

The ALSA API allows a PCM to return multiple descriptors from `snd_pcm_poll_descriptors`, and then the `snd_pcm_poll_descriptors_revents`  call is used to turn this into a suitable POLLIN/POLLOUT response (known in the ALSA API as *demangling*). My PCM returns at least two descriptors for each call to `snd_pcm_poll_descriptors`, and expects to have the same descriptors (in the same order) returned to `snd_pcm_poll_descriptors_revents` so that the PCM can work out what the poll result was. As far as I can tell this is what a client of the ALSA API [is supposed to do](https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#ga7e561f305702c6f52dab49b6c84f7df7).

MPD, however, provides only a single poll descriptor to `snd_pcm_poll_descriptors_revents`, which at the moment causes my PCM to return `-EINVAL`. This obviously causes MPD to cease playback. The line of code in question is:

https://github.com/MusicPlayerDaemon/MPD/blob/a7ba10423df80dbbe2e5959c5135f899f454fd5d/src/lib/alsa/NonBlock.cxx#L72

I'm aware that returning multiple fds for a PCM is unusual, but it is legal. I'm going to try to work around MPD's interpretation of the ALSA API, but I felt that I should also raise this bug as I think that MPD's behaviour is incorrect in this case.

**Comment by MaxKellermann** [comment]
Duplicate of https://github.com/MusicPlayerDaemon/MPD/issues/432
See https://github.com/MusicPlayerDaemon/MPD/issues/432#issuecomment-446350833 and https://github.com/MusicPlayerDaemon/MPD/issues/432#issuecomment-456174612 for details.
TLDR: your expectation may sound reasonable, but not documented, thus wrong.

**Comment by timothyjward** [comment]
> Duplicate of #432
> See [#432 (comment)](https://github.com/MusicPlayerDaemon/MPD/issues/432#issuecomment-446350833) and [#432 (comment)](https://github.com/MusicPlayerDaemon/MPD/issues/432#issuecomment-456174612) for details.
> TLDR: your expectation may sound reasonable, but not documented, thus wrong.

For reference, the internals of the ALSA library also have this expectation. The `hw` PCM code provides two descriptors and expects to receive both, in the same order, when calling to `snd_pcm_poll_descriptors_revents`. Anything else results in `-EINVAL` from the PCM. The code is visible here:

https://github.com/alsa-project/alsa-lib/blob/23a191a82c693456e61431ab699cddc1e5782a26/src/pcm/pcm_hw.c#L222-L254

I agree that the documentation is poor.

**Comment by MaxKellermann** [comment]
I know this piece of code. And it's buggy according to the API documentation.
I repeat what I said before several times: as soon as the API documentation gets updated, I will change MPD according to the new API.
But that's not the current API. MPD is correct, and that piece of code is buggy.
